de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,543 @@
+
+        if (referencePoint == null) {
+            LOG.warn("could not remove referencePoint, no referencePoint is registerid with ID: "
+                + id);
+            return;
+        }
+
+        if (partiallySharedReferencePoints.contains(referencePoint))
+            partiallySharedReferencePoints.remove(referencePoint);
+        else
+            completelySharedReferencePoints.remove(referencePoint);
+
+        idToReferencePointMapping.remove(id);
+        referencePointToIDMapping.remove(referencePoint);
+        partiallySharedResourceMapping.remove(referencePoint);
+
+        LOG.debug("removed referencePoint " + referencePoint + " with ID " + id);
+    }
+
+    /**
+     * Adds the given resources to a <b>partially</b> shared reference point.
+     * 
+     * @param referencePoint
+     *            a referencePoint that was added as a partially shared
+     *            referencePoint
+     * @param resources
+     *            the resources to add
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
+     * 
+     * @throws IllegalStateException if the project is completely or not shared
+     * at all
+     */
+    public synchronized void addResources(IReferencePoint referencePoint,
+        Collection<? extends IResource> resources) {
+
+        if (referencePointToIDMapping.get(referencePoint) == null) {
+            LOG.warn("could not add resources to referencePoint "
+                + referencePoint + " because it is not shared");
+            // throw new IllegalStateException(
+            // "could not add resources to project " + project
+            // + " because it is not shared");
+            return;
+        }
+
+        if (completelySharedReferencePoints.contains(referencePoint)) {
+            LOG.warn("cannot add resources to completely shared referencePoint: "
+                + referencePoint);
+            // throw new IllegalStateException(
+            // "cannot add resources to completely shared project: " + project);
+            return;
+        }
+
+        Set<IResource> partiallySharedResources = partiallySharedResourceMapping
+            .get(referencePoint);
+
+        if (partiallySharedResources.isEmpty()) {
+            partiallySharedResources = new HashSet<IResource>(Math.max(1024,
+                (resources.size() * 3) / 2));
+
+            partiallySharedResourceMapping.put(referencePoint,
+                partiallySharedResources);
+        }
+
+        partiallySharedResources.addAll(resources);
+    }
+
+    /**
+     * Removes the given resources from a <b>partially</b> shared reference
+     * point.
+     * 
+     * @param referencePoint
+     *            a reference point that was added as a partially shared
+     *            reference point
+     * @param resources
+     *            the resources to remove
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=210&issues=AWVWjH4plRPS_ea-bThW&open=AWVWjH4plRPS_ea-bThW)

[ISSUE_KEY]: AWVWjH4plRPS_ea-bThW